README.md

```diff
@@ -130,343 +135,328 @@ ag https://raw.githubusercontent.com/asyncapi/generator/v1.0.1/test/docs/dummy.y
+ * Check out package.json file and look for `generator.parameters.maxTextSize` and its description and default value.
  */
-maxTextSize: {{ params.maxTextSize }}
+function Scripts({ params }) {
+  return `
+<script src="https://cdn.jsdelivr.net/npm/mermaid/dist/mermaid.min.js"></script>
+<script>
+  mermaid.initialize({
+    startOnLoad: true,
+    maxTextSize: ${params.maxTextSize},
+  });
+</script>
+`;
+}
 ```
 
-#### Using Filters
+### Custom (reusable) components
 
-We can differentiate three types of filters in the Generator:
-- default filters that are part of the Nunjucks templating engine
-- generic filters provided by AsyncAPI Initiative
-- your custom template filters
+Check out [template/index.js](template/index.js) file to see an example how you can create reusable components and use them inside template:
 
-##### Nunjucks Built-In Filters
+```js
+// Import custom components from file 
+import { HTML, Head, Body } from "../partials/common";
+
+/* 
+ * Below you can see how reusable chunks (components) could be called.
+ * Just write a new component (or import it) and place it inside the File or another component.
+ * 
+ * Notice that you can pass parameters to components. In fact, underneath, each component is a pure Javascript function.
+ */
+export default function({ asyncapi, params }) {
+  ...
+  const cssLinks = [
+    'https://unpkg.com/tailwindcss@^1.0/dist/tailwind.min.css',
+    'style.css',
+  ];
+
+  // Notice that root component is the `File` component.
+  return (
+    <File name="index.html">
+      <HTML>
+        <Head cssLinks={cssLinks} />
+        <Body>
+          <BodyContent asyncapi={asyncapi} />
+          ...
+        </Body>
+      </HTML>
+    </File>
+  );
+}
 
-All built-in filters are listed [here](https://mozilla.github.io/nunjucks/templating.html#builtin-filters).
+// Custom component inside template file
+function BodyContent({ asyncapi }) {
+  // implemntation...
+}
+```
 
-The most commonly used filter during template development is `dump` that internally calls `JSON.stringify` and allows you to render in your template file the entire object. Very useful during debugging. 
+Each custom component must returns as output pure `string`, another custom component, `null` or `undefined`. Nothing will be rendered for the last two.
 
-In [partials/diagramContent.html](partials/diagramContent.html) file there is a use case for `dictsort` filter:
-```
-<!-- Usage of build-in "dictsort" Nunjucks filter for sorting items alphabetically 
-and how it works in a for loop -->
-{% for schemaName in asyncapi.components().schemas() | dictsort %}
-    <li>{{ schemaName[0] }}</li>
-{% endfor %}
-```
+Recommended place to create reusable chunks is the `partials` folder. The reusable parts can be located both in the `template` folder as in another named folder. The only exception is the `hooks` folder, it is reserved for the Generator.
```

Comment:
really `partials`? it is related to nunjucks, we should promote that components go to `template/components` and other functions to `template/helpers` and therefore all template related goes to `template`

Reply:
I thought about the components folder, but this name is too much associated with UI and not templates, but helpers is a good idea. Of course, these are just developer recommendations and as it says, people can create components in the templates folder as well. React is not restrictive as Nunjucks.

Reply:
awesome, so please move them to templates folder under some subdirectory, I'm ok with components but if you have strong feelings against then choose something else.

the only problem I see is that in markdown template you have all in partials, so we will have it a bit inconsistent, we should choose one recommendation here that will be used everywhere, at least in asyncapi org

Reply:
regarding the agreement, I [see](https://github.com/asyncapi/template-for-generator-templates/pull/6#discussion_r539943292) that for example, that @fmvilas  prefers to have stuff not in `partials` but outside `template`. We should agree on one

cc @jonaslagoni

Reply:
As you want, we can rename to `components` and create fast PR with renaming also in `markdown-template`.

Reply:
Hmm, I agree we should promote a standard folder to place partials in (so people can easily get started without having to take an active choice). For me I dont really like having the `template` folder used for partials since it should reassembly the actual generated files as close as possible and partials would clutter this.

But I have no strong opinion whether we suggest a `helpers`, `partials` or `darth wader` folder for storing them in the root folder of the template 😄

Reply:
IMHO, I think it should be `components` and it should be outside the `template` folder. The reason is what @jonaslagoni said, to make it clear what files are meant to be part of the output and what aren't.

That said, what was the reason to have a `template` folder? IIRC, it was to differentiate the `package.json` of the template from a potentially generated `package.json`. Same for `README.md`, `package-lock.json`, etc. Thinking about that, I think our structure should be:

```
template
  ... // potential output files
components
  ... // reusable parts (same as partial, conceptually)
package.json
README.md
etc.
```

If you look at that, it's the same structure Next.js uses, they just call it `pages` instead of `template`:

```
pages
  ... // available pages
components
  ... // reusable parts
package.json
README.md
etc.
```

Therefore, it's not a strange pattern. I'm also happy with `files` instead of `template` if that makes more sense.